File: gg_arch/boardgame.py
# Copyright 2013 Hans Rinderknecht
#
# This file is part of ggPlayer.
# ggPlayer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ggPlayer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ggPlayer.  If not, see <http://www.gnu.org/licenses/>.

class Boardgame(object):
    """Common boardgame class

    The basic class for a boardgame object.
    Specific games (including boards, pieces, and rulesets) will be created as subclasses of this object.

    Attributes:
        board (dict):  Game board definition.  board[cell.name]: gameCell object
        state (dict):  Current state definition.  state[cell.name]: list of gamepiece.ID currently in cell
        history (list):  History of executed moves, appended from end
        pieces (list):  List of Gamepiece objects.  pieces[gamepiece.ID] = gamepiece 
        gamename (str):  Name of the game (for reference only)
        boardtype (str):  Name of the board type (for reference only)
        num_players (int):  Number of players
    """

    board = {}  
    state = {}  
    history = [] 
    pieces = []  
    gamename = ''
    boardtype = '' 
    num_players = 0 

    def __init__(self, gamename):
        """Constructor for the boardgame object

        :param gamename:  The name of the game
        :type gamename:  str

        """
        self.gamename = gamename

    def getBoard(self):
        """Return the current state of the board"""
        return self.board

    def getHistory(self):
        """Return the history of executed moves"""
        return self.history

    def make_move(self, piece, cell):
        """Check if a proposed move is legal; modify the game state and history

        :param piece: The piece to be moved.
        :type piece: :class:Gamepiece object
        :param cell:  The destination cell of the piece.
        :type cell: :class:gameCell object
        :returns:  bool -- true, if successful
            
        """
        # make_move goes through the following steps:
        # 0. Verify piece_ID is legal to be moved this turn (belongs to Player)
        # 1. Calculate all legal cells for piece_ID; confirm cell_ID is one of the legal cells
        # 2. Check resulting state of gameboard to verify it is legal (e.g. not putting self into check)
        # 3. Trigger actions associated with the move (kill attacked piece, promote pawn, switch to other player's turn, etc)
        # 4. Return true when complete; return false, otherwise.
        #Window.alert("called")
        flag = True
        if self.isLegalMove(piece, cell):
            testState = self.newState(piece, cell)
            legal, event, args = self.isLegalState(testState)
            if legal:
                self.state = testState
                self.pieces[piece.ID].location = cell.name
                movecode = piece.name + cell.name
                self.history.append(movecode)
                if event!=[]:
                    self.doEvent(event, args)
            else: flag = False
        else: flag = False

        return flag

    def newState(self, piece, cell):
        """Generic new state from moving piece to cell

        :param piece: The piece to be moved.
        :type piece: :class:Gamepiece object
        :param cell:  The destination cell of the piece.
        :type cell: :class:gameCell object
        :returns:  newState, a state for this game
            
        """
        newState = self.state
        newState[piece.location].remove(piece.ID)
        newState[cell.name].append(piece.ID)
        return newState


    def isLegalMove(self, piece, cell):
        """Determine if moving piece to cell is a legal move"""
        legalMoves = self.getLegalMoves(piece)
        if cell.name in legalMoves:
            return True
        else:
            return False

    def getLegalMoves(self, piece):
        """Generic list of legal moves for a piece"""
        return self.board.keys()

    def isLegalState(teststate):
        """Generic determine if board state is legal"""
        return True, [], []

    def doEvent(self, event, args):
        """Generic execute an event precipitated by a move (e.g. king in check, kill piece, ..."""
        return True

File: gg_arch/boardgame.js
/* start module: boardgame */
$pyjs.loaded_modules['boardgame'] = function (__mod_name__) {
	if($pyjs.loaded_modules['boardgame'].__was_initialized__) return $pyjs.loaded_modules['boardgame'];
	var $m = $pyjs.loaded_modules["boardgame"];
	$m.__repr__ = function() { return "<module: boardgame>"; };
	$m.__was_initialized__ = true;
	if ((__mod_name__ === null) || (typeof __mod_name__ == 'undefined')) __mod_name__ = 'boardgame';
	$m.__name__ = __mod_name__;


	$m['Boardgame'] = (function(){
		var $cls_definition = new Object();
		var $method;
		$cls_definition.__module__ = 'boardgame';
		$cls_definition['board'] = $p['dict']([]);
		$cls_definition['state'] = $p['dict']([]);
		$cls_definition['history'] = $p['list']([]);
		$cls_definition['pieces'] = $p['list']([]);
		$cls_definition['gamename'] = '';
		$cls_definition['boardtype'] = '';
		$cls_definition['num_players'] = 0;
		$method = $pyjs__bind_method2('__init__', function(gamename) {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
				gamename = arguments[1];
			}

			self.gamename = gamename;
			return null;
		}
	, 1, [null,null,['self'],['gamename']]);
		$cls_definition['__init__'] = $method;
		$method = $pyjs__bind_method2('getBoard', function() {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
			}

			return $p['getattr'](self, 'board');
		}
	, 1, [null,null,['self']]);
		$cls_definition['getBoard'] = $method;
		$method = $pyjs__bind_method2('getHistory', function() {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
			}

			return $p['getattr'](self, 'history');
		}
	, 1, [null,null,['self']]);
		$cls_definition['getHistory'] = $method;
		$method = $pyjs__bind_method2('make_move', function(piece, cell) {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
				piece = arguments[1];
				cell = arguments[2];
			}
			var $add2,args,movecode,legal,flag,$add1,testState,event;
			flag = true;
			if ($p['bool'](self['isLegalMove'](piece, cell))) {
				testState = self['newState'](piece, cell);
				var $tupleassign1 = $p['__ass_unpack'](self['isLegalState'](testState), 3, null);
				legal = $tupleassign1[0];
				event = $tupleassign1[1];
				args = $tupleassign1[2];
				if ($p['bool'](legal)) {
					self.state = testState;
					$p['getattr'](self, 'pieces').__getitem__($p['getattr'](piece, 'ID')).location = $p['getattr'](cell, '$$name');
					movecode = $p['__op_add']($add1=$p['getattr'](piece, '$$name'),$add2=$p['getattr'](cell, '$$name'));
					self['history']['append'](movecode);
					if ($p['bool'](!$p['op_eq'](event, $p['list']([])))) {
						self['doEvent'](event, args);
					}
				}
				else {
					flag = false;
				}
			}
			else {
				flag = false;
			}
			return flag;
		}
	, 1, [null,null,['self'],['piece'],['cell']]);
		$cls_definition['make_move'] = $method;
		$method = $pyjs__bind_method2('newState', function(piece, cell) {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
				piece = arguments[1];
				cell = arguments[2];
			}
			var newState;
			newState = $p['getattr'](self, 'state');
			newState.__getitem__($p['getattr'](piece, 'location'))['remove']($p['getattr'](piece, 'ID'));
			newState.__getitem__($p['getattr'](cell, '$$name'))['append']($p['getattr'](piece, 'ID'));
			return newState;
		}
	, 1, [null,null,['self'],['piece'],['cell']]);
		$cls_definition['newState'] = $method;
		$method = $pyjs__bind_method2('isLegalMove', function(piece, cell) {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
				piece = arguments[1];
				cell = arguments[2];
			}
			var legalMoves;
			legalMoves = self['getLegalMoves'](piece);
			if ($p['bool'](legalMoves.__contains__($p['getattr'](cell, '$$name')))) {
				return true;
			}
			else {
				return false;
			}
			return null;
		}
	, 1, [null,null,['self'],['piece'],['cell']]);
		$cls_definition['isLegalMove'] = $method;
		$method = $pyjs__bind_method2('getLegalMoves', function(piece) {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
				piece = arguments[1];
			}

			return self['board']['keys']();
		}
	, 1, [null,null,['self'],['piece']]);
		$cls_definition['getLegalMoves'] = $method;
		$method = $pyjs__bind_method2('isLegalState', function() {
			if (this.__is_instance__ === true) {
				var teststate = this;
			} else {
				var teststate = arguments[0];
			}

			return $p['tuple']([true, $p['list']([]), $p['list']([])]);
		}
	, 1, [null,null,['teststate']]);
		$cls_definition['isLegalState'] = $method;
		$method = $pyjs__bind_method2('doEvent', function(event, args) {
			if (this.__is_instance__ === true) {
				var self = this;
			} else {
				var self = arguments[0];
				event = arguments[1];
				args = arguments[2];
			}

			return true;
		}
	, 1, [null,null,['self'],['event'],['args']]);
		$cls_definition['doEvent'] = $method;
		var $bases = new Array($p['object']);
		var $data = $p['dict']();
		for (var $item in $cls_definition) { $data.__setitem__($item, $cls_definition[$item]); }
		return $p['_create_class']('Boardgame', $p['tuple']($bases), $data);
	})();
	return this;
}; /* end boardgame */


/* end module: boardgame */


